mpeg2/src/ts.rs

```diff
@@ -18,85 +20,205 @@ pub struct AdaptationField {
     pub discontinuity_indicator: Option<bool>,
     pub random_access_indicator: Option<bool>,
     pub program_clock_reference_27mhz: Option<u64>,
+    pub private_data_types: Vec<u8>,
+    pub temi_timeline_descriptors: Vec<TEMITimelineDescriptor>,
 }
 
 impl AdaptationField {
     pub fn decode(buf: &[u8]) -> Result<Self, DecodeError> {
+        let mut bs = Bitstream::new(buf);
         let mut af = Self::default();
-        let af_length = buf[0];
-        if af_length as usize > buf.len() - 1 {
+        let af_length = bs.read_u8();
+        if bs.remaining_bytes() < af_length as usize {
             return Err(DecodeError::new("adaptation field length too long"));
-        } else if buf[0] > 0 {
-            af.discontinuity_indicator = Some(buf[1] & 0x80 != 0);
-            af.random_access_indicator = Some(buf[1] & 0x40 != 0);
-            af.program_clock_reference_27mhz = if af_length >= 7 && buf[1] & 0x10 != 0 {
-                let base = (buf[2] as u64) << 25 | (buf[3] as u64) << 17 | (buf[4] as u64) << 9 | (buf[5] as u64) << 1 | (buf[6] as u64) >> 7;
-                let ext = ((buf[6] as u64) & 1) << 8 | (buf[7] as u64);
+        } else if af_length > 0 {
+            af.discontinuity_indicator = Some(bs.read_boolean());
+            af.random_access_indicator = Some(bs.read_boolean());
```

Comment:
What would you think about making the `read_` methods return `Result` types so they could bubble up an error if there aren't enough bytes remaining? We could give them a `name` parameter so they could still provide meaningful errors:

```suggestion
            af.discontinuity_indicator = Some(bs.read_boolean("discontinuity_indicator")?);
            af.random_access_indicator = Some(bs.read_boolean("random_access_indicator")?);
```

This would allow us to get rid of the length checks from this function entirely and be a much more foolproof way to make sure it never panics.

Reply:
I haven't got used to idiomatic way to handle errors in Rust. will update the PR.